File: backend/flask/app/templates/construction/test_construction.py
"""
Tests avancés module Construction (unitaires, intégration, e2e, sécurité, i18n, plugins).
"""
import pytest
from flask import Flask
from flask_jwt_extended import create_access_token
from backend.flask.app.templates.construction import construction_bp

@pytest.fixture
def client():
    app = Flask(__name__)
    app.config['JWT_SECRET_KEY'] = 'test'
    app.register_blueprint(construction_bp)
    # ... setup JWT, plugins, etc. ...
    with app.test_client() as client:
        yield client

def test_list_construction(client):
    token = create_access_token(identity={'role': 'admin'})
    rv = client.get('/api/construction/', headers={'Authorization': f'Bearer {token}'})
    assert rv.status_code == 200
    assert 'data' in rv.json

def test_construction_route_security(client):
    """Teste la sécurité de la route Construction (CORS, JWT, RBAC, audit)."""
    response = client.get('/construction')
    assert response.status_code in (200, 401, 403)

# ... autres tests : sécurité, plugins, i18n, audit, export, etc. ...

File: frontend/src/tests/integration/construction/test_construction.js
/**
 * Test d'intégration avancé pour la gestion des projets Construction (REST, GraphQL, sécurité, i18n, plugins, RGPD, audit, multitenancy, fallback IA, SEO, multilingue)
 * @see README.md pour la documentation complète
 */
const request = require('supertest');
const app = require('../../../../backend/node/app');
const { getJwt, SUPPORTED_LANGUAGES } = require('../../utils/test_utils');

describe('Construction API Integration', () => {
  let jwt;
  beforeAll(async () => {
    jwt = await getJwt('admin');
  });
  it('GET /construction/projects (secured, i18n, audit, plugins)', async () => {
    for (const lang of SUPPORTED_LANGUAGES) {
      const res = await request(app)
        .get('/construction/projects')
        .set('Authorization', `Bearer ${jwt}`)
        .set('Accept-Language', lang);
      expect(res.statusCode).toBe(200);
      expect(res.body).toHaveProperty('data');
      expect(res.headers['content-language']).toBe(lang);
    }
  });
  it('POST /construction/projects (validation, RGPD, plugins, audit)', async () => {
    const res = await request(app)
      .post('/construction/projects')
      .set('Authorization', `Bearer ${jwt}`)
      .send({ name: 'Chantier X', type: 'BTP', owner: 'user_id' });
    expect(res.statusCode).toBe(201);
    expect(res.body).toHaveProperty('project');
  });
  it('GraphQL /construction/graphql (roles, plugins, fallback IA)', async () => {
    const query = '{ projects { id name type owner } }';
    const res = await request(app)
      .post('/construction/graphql')
      .set('Authorization', `Bearer ${jwt}`)
      .send({ query });
    expect(res.statusCode).toBe(200);
    expect(res.body.data).toHaveProperty('projects');
  });
  it('Sécurité CORS, WAF, anti-DDOS', async () => {
    const res = await request(app)
      .options('/construction/projects')
      .set('Origin', 'https://evil.com');
    expect(res.headers).toHaveProperty('access-control-allow-origin');
  });
});
